forge/src/main/java/org/valkyrienskies/mod/forge/mixin/compat/cc_tweaked/MixinTurtleBrain.java

```diff
@@ -0,0 +1,125 @@
+package org.valkyrienskies.mod.forge.mixin.compat.cc_tweaked;
+
+import dan200.computercraft.shared.turtle.blocks.TileTurtle;
+import dan200.computercraft.shared.turtle.core.TurtleBrain;
+import java.util.List;
+import javax.annotation.Nonnull;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.core.Vec3i;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import org.joml.Matrix4dc;
+import org.joml.Vector3d;
+import org.joml.Vector3dc;
+import org.joml.primitives.AABBic;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Pseudo;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
+import org.valkyrienskies.core.api.Ship;
+import org.valkyrienskies.mod.common.VSGameUtilsKt;
+import org.valkyrienskies.mod.common.config.VSGameConfig;
+import org.valkyrienskies.mod.common.util.VectorConversionsMCKt;
+
+@Pseudo
+@Mixin(TurtleBrain.class)
+public abstract class MixinTurtleBrain {
+    @Shadow
+    public abstract TileTurtle getOwner();
+
+    @Shadow
+    public abstract void setOwner(TileTurtle owner);
+
+    @Shadow
+    @Nonnull
+    public abstract Level getWorld();
+
+    @ModifyVariable(
+        method = "Ldan200/computercraft/shared/turtle/core/TurtleBrain;teleportTo(Lnet/minecraft/world/level/Level;Lnet/minecraft/core/BlockPos;)Z",
+        at = @At(value = "HEAD"),
+        index = 2
+    )
+    private BlockPos ValkyrienSkies2$teleportToBlockPos(BlockPos pos) {
+        TileTurtle currentOwner = getOwner();
+        BlockPos oldPos = currentOwner.getBlockPos();
+        Level world = getWorld();
+
+        Ship ship = getShip(world, oldPos);
+        if (ship != null) {
+            // THERE IS A SHIP
+            Direction d = getNewDirection(ship, currentOwner.getDirection());
+            if (!doesShipContainPoint(ship, pos)) {
+                // POSITION IS OUTSIDE THE SHIP'S AABB
+
+                currentOwner.setDirection(d);
+                setOwner(currentOwner);
+
+                if (!isShipScaled(ship)) {
+                    // SHIP IS NOT SCALED
+
+                    return getWorldPosFromShipPos(ship, pos);
+                } else if (turtlesLeaveScaledShips()) {
+                    // SHIP IS SCALED AND TURTLES CAN LEAVE SCALED SHIPS
+
+                    return getWorldPosFromShipPos(ship, pos);
+                } else {
+                    // SHIP IS SCALED AND TURTLES CAN NOT LEAVE SCALED SHIPS
+
+                    return oldPos;
+                }
+            }
+        }
+        return pos;
+    }
+
+    // CUSTOM METHODS
+
+    private static Direction getNewDirection(Ship ship, Direction direction) {
```

Comment:
Can you make sure to `@Unique` all of these helper functions?

Reply:
Sure